harmony_netcdf_to_zarr/mosaic_utilities.py

```diff
@@ -89,46 +164,82 @@ def _parse_variable_dimensions(self, variable: Variable, dataset: Dataset):
                 dim_data = self.input_dimensions.setdefault(dimension_path, {})
 
                 if input_path not in dim_data:
-                    dim_data[input_path] = DimensionInformation(dataset,
-                                                                dimension_path)
+                    dim_data[input_path] = NetCDF4DimensionInformation(
+                        dataset, dimension_path
+                    )
 
+    def _aggregate_output_dimensions(self):
+        """ Iterate through each input dimension listed in
+            `self.input_dimensions`. For each combine the values from all
+            input granules into a single array. Derive an output grid that is
+            evenly spaced and has points at all the input values.
 
-class DimensionInformation:
-    """ A class containing information on a dimension, including the path of
-        the dimension variable within a NetCDF-4 or Zarr Dataset, the values of
-        the 1-dimensional dimension array and any associated temporal epoch
-        and unit. For TRT-121, it is initially assumed that non-temporal
-        dimensions can be aggregated without requiring offsets similar to
-        that stored in the temporal dimensions `units` metdata attribute.
-
-        This class can be used to contain both the dimension information from
-        individual input NetCDF-4 files, as well as the aggregated output
-        dimension.
-
-    """
-    def __init__(self, dataset: Dataset, dimension_path: str):
-        self.dimension_path = dimension_path
-        self.values = dataset[dimension_path][:]
-        self.units = get_nc_attribute(dataset[dimension_path], 'units')
-        self.epoch = None
-        self.time_unit = None
-        self._get_epoch_and_unit()
-
-    def _get_epoch_and_unit(self):
-        """ Check the `units` attribute in the dimension variable metadata. If
-            present, compare the format to the CF-Convention format for a
-            temporal dimension (e.g., "seconds since 2000-01-02T03:04:05"), and
-            extract the type of unit.
-
-            For documentation on temporal dimensions see:
+        """
+        for dimension_name, dimension_inputs in self.input_dimensions.items():
+            are_inputs_temporal = [dimension_input.is_temporal()
+                                   for dimension_input
+                                   in dimension_inputs.values()]
+
+            if all(are_inputs_temporal):
+                self.output_dimensions[dimension_name] = self._get_temporal_output_dimension(
+                    dimension_inputs, dimension_name
+                )
+            elif any(are_inputs_temporal):
+                raise MixedDimensionTypeError(dimension_name)
+            else:
+                all_input_values = np.unique(
```

Comment:
-If more than one value for a time-stamp exists, do we want an exception here?  Mosaicing with interpolation over multiple values or near-by values is wholly separate from this implementation.-  NM - this is the else branch for non-temporal values.  So this code is mosaicing (non-interpolated) across non-temporal dimensions?  (Perhaps a comment for the limited in world-view).

Reply:
Yes - this branch is for non-temporal dimensions. I will add comments in each of the branches of the if-statement to clarify.

Generally, though, my approach in this PR is: create aggregated dimensions for _all_ input dimensions (including non-temporal). If we only want to use the output temporal dimensions in the output Zarr object, then we can determine that when forming the output. This class is more about the gathering of aggregated values, rather than deciding whether to use them. (And should be a good foundation for spatial aggregation in the future)

Reply:
I've added a comment to this branch of the if statement and to the all-temporal branch to clarify the conditions and why they are required.

Reply:
Actually, my comment here is not dimension-type specific or time-dimension specific, it is that our code is really only prepared to aggregate across non-overlapping grid dimension-scales.  A non-unique situation is essentially an invalid dimension-scale.  Again, wondering if an exception is warranted here.

Reply:
Also - we're probably ok assuming valid dimension scales, and the relatively simplistic concatenate operation.  Perhaps some exception in trying to write the data will clue us into invalid dimension scales.  I'm tempted to verify consistency and validity before we we encounter failures, but it is hard to balance cost of coding vs. likelihood of errors.  In established datasets - very low likelihood of errors.  In new data products, however, we're debugging while the science team is debugging, and sometimes such checks in our code is worth it.